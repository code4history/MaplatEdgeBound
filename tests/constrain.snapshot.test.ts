// edge-ruler.snapshot.test.ts
/**
 * このテストファイルは状態変化のスナップショットを記録し、
 * システム全体の振る舞いを検証することを目的としています。
 * 
 * ユニットテストは edge-ruler.unit.test.ts にあります。
 */
import { describe, it, expect, beforeEach } from 'vitest';
import EdgeRuler from '../src/variant/constrain';
import Delaunator from 'delaunator';

// 型定義を追加
interface EdgeState {
  operation: string;
  edge01: number;
  edge02: number;
}

interface OperationResult {
  success: boolean;
  result?: unknown;
  error?: string;
}

describe('EdgeRuler Integration-Based Unit Tests', () => {
  const samplePoints = [[150, 50], [50, 200], [150, 350], [250, 200]];
  
  describe('State Transitions', () => {
    let ruler: EdgeRuler;
    let del: Delaunator<Float64Array>;
    
    beforeEach(() => {
      del = Delaunator.from(samplePoints);
      ruler = new EdgeRuler(del);
    });

    it('should record state changes during constraint operation', () => {
      /*const initialState = {
        triangles: Array.from(ruler.del.triangles),
        halfedges: Array.from(ruler.del.halfedges),
        vertMap: Array.from(ruler.vertMap)
      };

      const edge = ruler.constrainOne(0, 2);*/

      const afterConstraint = {
        triangles: Array.from(ruler.del.triangles),
        halfedges: Array.from(ruler.del.halfedges),
        vertMap: Array.from(ruler.vertMap)
      };

      expect(afterConstraint).toMatchSnapshot('state-after-constraint');
    });

    it('should maintain correct edge states through operations', () => {
      // 型付きの配列を初期化
      const transitions: EdgeState[] = [];

      transitions.push({
        operation: 'initial',
        edge01: ruler.findEdge(0, 1),
        edge02: ruler.findEdge(0, 2)
      });

      ruler.constrainOne(0, 2);
      transitions.push({
        operation: 'constrain-0-2',
        edge01: ruler.findEdge(0, 1),
        edge02: ruler.findEdge(0, 2)
      });

      ruler.delaunify(true);
      transitions.push({
        operation: 'delaunify',
        edge01: ruler.findEdge(0, 1),
        edge02: ruler.findEdge(0, 2)
      });

      expect(transitions).toMatchSnapshot('edge-state-transitions');
    });
  });

  describe('Complex Operations', () => {
    it('should handle sequence of operations from integration test', () => {
      const del = Delaunator.from(samplePoints);
      const ruler = new EdgeRuler(del);

      const operations = [
        () => ruler.constrainOne(0, 2),
        () => ruler.delaunify(true),
        () => ruler.findEdge(0, 2),
        () => ruler.isConstrained(ruler.findEdge(0, 2))
      ];

      const results: OperationResult[] = operations.map(op => {
        try {
          return {
            success: true,
            result: op()
          };
        } catch (e) {
          if (e instanceof Error) {
            return {
              success: false,
              error: e.message
            };
          }
          return {
            success: false,
            error: 'Unknown error'
          };
        }
      });

      expect(results).toMatchSnapshot('operation-sequence-results');
    });
  });

  describe('Edge Cases', () => {
    it('should handle almost-collinear points correctly', () => {
      // 完全な一直線ではなく、わずかにずれた点を使用
      const almostCollinearPoints = [
        [0, 0], 
        [1, 0.001], 
        [2, -0.001], 
        [3, 0.002]
      ];
      const del = Delaunator.from(almostCollinearPoints);
      const ruler = new EdgeRuler(del);
  
      const result = ruler.constrainOne(0, 2);
      expect(result).toMatchSnapshot('almost-collinear-constraint-result');
    });
  
    it('should handle near-but-valid points', () => {
      // 近接しているが、有効な三角形分割が可能な点を使用
      const nearButValidPoints = [
        [0, 0], 
        [0.1, 0.1], 
        [1, 0], 
        [1, 1]
      ];
      const del = Delaunator.from(nearButValidPoints);
      const ruler = new EdgeRuler(del);
  
      const result = ruler.constrainOne(0, 2);
      expect(result).toMatchSnapshot('near-valid-constraint-result');
    });
  
    // エラーケースのテストを追加
    it('should throw error for true collinear points', () => {
      const collinearPoints = [[0, 0], [1, 1], [2, 2], [3, 3]];
      const del = Delaunator.from(collinearPoints);
      
      expect(() => new EdgeRuler(del))
        .toThrow('No edges in triangulation');
    });
  });
});