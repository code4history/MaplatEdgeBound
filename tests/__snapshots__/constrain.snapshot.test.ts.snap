// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EdgeBound Integration-Based Unit Tests > Complex Operations > should handle sequence of operations from integration test > operation-sequence-results 1`] = `
[
  {
    "result": 2,
    "success": true,
  },
  {
    "result": Constrain {
      "consd": BitSet8 {
        "W": 8,
        "bs": Uint8Array [
          20,
        ],
      },
      "del": Delaunator {
        "_cx": 150,
        "_cy": 158.33333333333331,
        "_dists": Float64Array [
          11736.111111111108,
          11736.111111111113,
          36736.11111111112,
          11736.111111111113,
        ],
        "_halfedges": Int32Array [
          -1,
          -1,
          4,
          -1,
          2,
          -1,
          0,
          0,
          0,
        ],
        "_hashSize": 2,
        "_hullHash": Int32Array [
          0,
          1,
        ],
        "_hullNext": Uint32Array [
          1,
          2,
          3,
          0,
        ],
        "_hullPrev": Uint32Array [
          3,
          0,
          1,
          2,
        ],
        "_hullStart": 1,
        "_hullTri": Uint32Array [
          2,
          3,
          4,
          1,
        ],
        "_ids": Uint32Array [
          0,
          1,
          3,
          2,
        ],
        "_triangles": Uint32Array [
          2,
          3,
          0,
          1,
          2,
          0,
          0,
          0,
          0,
        ],
        "coords": Float64Array [
          150,
          50,
          50,
          200,
          150,
          350,
          250,
          200,
        ],
        "halfedges": Int32Array [
          -1,
          -1,
          4,
          -1,
          2,
          -1,
        ],
        "hull": Uint32Array [
          1,
          2,
          3,
          0,
        ],
        "triangles": Uint32Array [
          2,
          3,
          0,
          1,
          2,
          0,
        ],
        "trianglesLen": 6,
      },
      "flips": BitSet8 {
        "W": 8,
        "bs": Uint8Array [
          0,
        ],
      },
      "vertMap": Uint32Array [
        1,
        5,
        3,
        0,
      ],
    },
    "success": true,
  },
  {
    "result": 2,
    "success": true,
  },
  {
    "result": true,
    "success": true,
  },
]
`;

exports[`EdgeBound Integration-Based Unit Tests > Edge Cases > should handle almost-collinear points correctly > almost-collinear-constraint-result 1`] = `-1`;

exports[`EdgeBound Integration-Based Unit Tests > Edge Cases > should handle near-but-valid points > near-valid-constraint-result 1`] = `-1`;

exports[`EdgeBound Integration-Based Unit Tests > State Transitions > should maintain correct edge states through operations > edge-state-transitions 1`] = `
[
  {
    "edge01": 2,
    "edge02": Infinity,
    "operation": "initial",
  },
  {
    "edge01": 5,
    "edge02": 2,
    "operation": "constrain-0-2",
  },
  {
    "edge01": 5,
    "edge02": 2,
    "operation": "delaunify",
  },
]
`;

exports[`EdgeBound Integration-Based Unit Tests > State Transitions > should record state changes during constraint operation > state-after-constraint 1`] = `
{
  "halfedges": [
    5,
    -1,
    -1,
    -1,
    -1,
    0,
  ],
  "triangles": [
    1,
    3,
    0,
    1,
    2,
    3,
  ],
  "vertMap": [
    1,
    2,
    3,
    4,
  ],
}
`;
